File: app/src/main/java/com/classicsviewer/app/utils/PatternBasedNormalizer.kt
package com.classicsviewer.app.utils

import android.util.Log
import com.classicsviewer.app.database.entities.NormalizationPatternEntity
import java.text.Normalizer
import java.util.concurrent.ConcurrentHashMap

/**
 * Data-driven text normalizer that applies regex patterns from the database.
 *
 * NOTE: This normalizer is NOT used for Greek or Latin.
 * - Greek uses GreekNormalizer (hardcoded)
 * - Latin has no normalization
 * - Other languages (Hebrew, Arabic, etc.) use this normalizer
 *
 * Normalization happens in two phases:
 * 1. NFD (Unicode Canonical Decomposition) - separates base chars from combining marks
 * 2. Apply custom regex patterns in priority order
 *
 * Example (Hebrew):
 *   Input: "דָּבָר" (with nikud)
 *   After NFD: "דָּבָר" (nikud separated as combining marks)
 *   After pattern [\u0591-\u05C7] → "": "דבר" (nikud removed)
 */
object PatternBasedNormalizer {
    private const val TAG = "PatternNormalizer"

    // Cache compiled regex patterns per language for performance
    private data class CompiledPattern(
        val regex: Regex,
        val replacement: String,
        val priority: Int,
        val description: String?
    )

    private val compiledCache = ConcurrentHashMap<String, List<CompiledPattern>>()

    /**
     * Normalize text using database-driven patterns for the given language.
     *
     * @param text The text to normalize
     * @param language The language code (e.g., "hebrew", "arabic") - NOT "greek" or "latin"
     * @param patterns The normalization patterns to apply (fetched from database)
     * @return Normalized text
     */
    fun normalize(text: String, language: String, patterns: List<NormalizationPatternEntity>): String {
        if (text.isEmpty()) return text

        val compiledPatterns = getCompiledPatterns(language, patterns)

        if (compiledPatterns.isEmpty()) {
            // No normalization rules for this language - return as-is
            return text
        }

        var result = text
        Log.d(TAG, "[$language] Input: '$result' (${result.length} chars)")

        // Step 1: Apply NFD normalization (separates base chars from diacritics)
        result = Normalizer.normalize(result, Normalizer.Form.NFD)
        Log.d(TAG, "[$language] After NFD: '$result' (${result.length} chars)")

        // Step 2: Apply custom regex patterns in priority order
        for ((index, compiled) in compiledPatterns.withIndex()) {
            try {
                val before = result
                result = compiled.regex.replace(result, compiled.replacement)
                if (before != result) {
                    Log.d(TAG, "[$language] Pattern ${index + 1}: '$before' -> '$result' (pattern: ${compiled.regex.pattern})")
                }
            } catch (e: Exception) {
                Log.w(TAG, "Error applying pattern '${compiled.regex.pattern}' for $language: ${e.message}")
            }
        }

        Log.d(TAG, "[$language] Final result: '$result' (${result.length} chars)")
        return result
    }

    /**
     * Convenience method that takes a single pattern (for testing)
     */
    fun normalize(text: String, language: String, pattern: NormalizationPatternEntity): String {
        return normalize(text, language, listOf(pattern))
    }

    /**
     * Get compiled regex patterns for a language, using cache for performance.
     */
    private fun getCompiledPatterns(language: String, patterns: List<NormalizationPatternEntity>): List<CompiledPattern> {
        // Create cache key from pattern contents (so cache updates if patterns change)
        val cacheKey = "$language:${patterns.size}:${patterns.hashCode()}"

        return compiledCache.getOrPut(cacheKey) {
            patterns
                .sortedBy { it.priority }  // Apply in priority order
                .mapNotNull { pattern ->
                    try {
                        CompiledPattern(
                            regex = Regex(pattern.pattern),
                            replacement = pattern.replacement,
                            priority = pattern.priority,
                            description = pattern.description
                        )
                    } catch (e: Exception) {
                        Log.w(TAG, "Failed to compile pattern '${pattern.pattern}' for $language: ${e.message}")
                        null
                    }
                }
        }
    }

    /**
     * Clear the compiled pattern cache (useful when patterns are updated)
     */
    fun clearCache() {
        compiledCache.clear()
        Log.d(TAG, "Normalization pattern cache cleared")
    }

    /**
     * Clear cache for a specific language
     */
    fun clearCache(language: String) {
        compiledCache.keys.removeIf { it.startsWith("$language:") }
        Log.d(TAG, "Normalization pattern cache cleared for $language")
    }
}

File: ios/ClassicsViewer/Utilities/PatternBasedNormalizer.swift
import Foundation

/// Data-driven text normalizer that applies regex patterns from database
/// Used for non-Greek/Latin languages (Hebrew, Arabic, Sanskrit, etc.)
/// Two-phase normalization: NFD decomposition + regex pattern application
class PatternBasedNormalizer {

    /// Compiled regex pattern with priority for sorting
    private struct CompiledPattern {
        let regex: NSRegularExpression
        let replacement: String
        let priority: Int
    }

    /// Cache of compiled patterns by language to avoid recompiling
    private static var patternCache: [String: [CompiledPattern]] = [:]
    private static let cacheLock = NSLock()

    /// Normalize text using database-driven regex patterns
    /// - Parameters:
    ///   - text: The text to normalize
    ///   - language: The language code (e.g., "hebrew", "arabic")
    ///   - patterns: Array of normalization patterns from database
    /// - Returns: Normalized text with patterns applied
    static func normalize(_ text: String, language: String, patterns: [NormalizationPattern]) -> String {
        guard !patterns.isEmpty else {
            return text.lowercased()
        }

        // Phase 1: NFD normalization (Unicode Canonical Decomposition)
        // This separates base characters from diacritics
        var result = text.decomposedStringWithCanonicalMapping

        // Phase 2: Apply regex patterns in priority order (highest priority first)
        let compiledPatterns = getCompiledPatterns(for: language, patterns: patterns)

        for compiled in compiledPatterns {
            let range = NSRange(result.startIndex..., in: result)
            result = compiled.regex.stringByReplacingMatches(
                in: result,
                options: [],
                range: range,
                withTemplate: compiled.replacement
            )
        }

        return result
    }

    /// Get compiled regex patterns from cache or compile them
    /// - Parameters:
    ///   - language: The language code for cache key
    ///   - patterns: The patterns to compile
    /// - Returns: Array of compiled patterns sorted by priority
    private static func getCompiledPatterns(for language: String, patterns: [NormalizationPattern]) -> [CompiledPattern] {
        cacheLock.lock()
        defer { cacheLock.unlock() }

        // Check cache first
        if let cached = patternCache[language] {
            return cached
        }

        // Compile patterns
        var compiled: [CompiledPattern] = []

        for pattern in patterns {
            do {
                let regex = try NSRegularExpression(pattern: pattern.pattern, options: [])
                compiled.append(CompiledPattern(
                    regex: regex,
                    replacement: pattern.replacement,
                    priority: pattern.priority
                ))
            } catch {
                print("Warning: Failed to compile regex pattern '\(pattern.pattern)' for \(language): \(error)")
                continue
            }
        }

        // Sort by priority (lowest number first - matches Android)
        compiled.sort { $0.priority < $1.priority }

        // Cache the compiled patterns
        patternCache[language] = compiled

        return compiled
    }

    /// Clear the pattern cache (useful when patterns are updated)
    /// - Parameter language: Optional language to clear, or nil to clear all
    static func clearCache(for language: String? = nil) {
        cacheLock.lock()
        defer { cacheLock.unlock() }

        if let lang = language {
            patternCache.removeValue(forKey: lang)
        } else {
            patternCache.removeAll()
        }
    }

    /// Normalize text using user-defined patterns (converts UserNormalizationPattern to NormalizationPattern)
    /// - Parameters:
    ///   - text: The text to normalize
    ///   - language: The language code
    ///   - userPatterns: Array of user normalization patterns
    /// - Returns: Normalized text
    static func normalize(_ text: String, language: String, userPatterns: [UserNormalizationPattern]) -> String {
        // Convert UserNormalizationPattern to NormalizationPattern
        let patterns = userPatterns.map { userPattern in
            NormalizationPattern(
                id: userPattern.id,
                language: userPattern.language,
                pattern: userPattern.pattern,
                replacement: userPattern.replacement,
                description: userPattern.description,
                priority: userPattern.priority
            )
        }

        return normalize(text, language: language, patterns: patterns)
    }
}
